bigsea-controller/service/api/controller/plugins/pid/pid_alarm.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (c) 2017 LSD - UFCG.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from utils.logger import ScalingLog
+import datetime
+import time
+
+class PIDAlarm:
+
+    ERROR_METRIC_NAME = "application-progress.error"
+    
+    def __init__(self, actuator, metric_source, trigger_down, trigger_up, min_cap, max_cap, 
+                            metric_rounding, heuristic_options, application_id, instances):
+        # TODO: Check parameters
+        self.metric_source = metric_source
+        self.actuator = actuator
+        self.trigger_down = trigger_down
+        self.trigger_up = trigger_up
+        self.min_cap = min_cap
+        self.max_cap = max_cap
+        self.metric_rounding = metric_rounding
+        self.heuristic_options = heuristic_options
+        self.application_id = application_id
+        self.instances = instances
+        self.integrated_error = 0
+
+        self.logger = ScalingLog("%s.pid.alarm.log" % (application_id), 
+                                 "controller.log", application_id)
+        self.cap_logger = ScalingLog("%s.cap.log" % (application_id), "cap.log", application_id)
+
+        self.last_error = ""        
+        self.last_progress_error_timestamp = datetime.datetime.strptime("0001-01-01T00:00:00.0Z", 
+                                                                        '%Y-%m-%dT%H:%M:%S.%fZ')
+        self.last_action = ""
+        self.cap = -1
+    
+    def check_application_state(self):
+        """
+            Checks the application progress by getting progress metrics from a 
+            metric source, checks if the metrics are new and tries to modify the
+            amount of allocated resources if necessary.
+        """
+        
+        # TODO: Check parameters
```

Comment:
Still **TODO**?

Reply:
Yes, I do not think it is going to be solved anytime soon.